students/Ruohan/lesson04/mailroom_part2.py

```diff
@@ -0,0 +1,115 @@
+#! /usr/bin/env python3
+
+import sys
+import string
+
+old_datas = [['William Gates III',  653784.39, 2],
+         ['Mark Zuckerberg', 16396.10, 3],
+         ['Jeff Bezos', 877.33, 1],
+         ['Paul Allen', 708.42, 3]]
+
+# make a new data structure for part2
+datas = []
+
+for i in old_datas:
+    dic = {}
+    dic.setdefault('name', i[0])
+    dic.setdefault('donation_amount', i[1])
+    dic.setdefault('donation_times',i[2])
+    dic.setdefault('recent_donation')
+    datas.append(dic)
+
+
+#make a donors' name list
+def donors_name_list():
+    donors = []
+    for data in datas:
+        donors.append(data['name'])
+    return donors
+
+
+def find_donor(donor,donors):
+    '''find if donor is already in the datas, if not add the name to the datas'''
+    if donor not in donors:
+        donor_info = {'name': donor}
+        donor_info.setdefault('donation_amount',0)
+        donor_info.setdefault('donation_times', 0)
+        datas.append(donor_info)
+    return donor
+
+
+def upgrade_datas(donor):
+    # upgrade donors' total donation amount, donation times and recent donation amount
+    amount = float(input('Enter donation amount: '))
+    for data in datas:
+        if donor == data.get('name'):
+            data['donation_amount'] +=  amount
+            data['donation_times'] += 1
+            data.setdefault('recent_donation', amount)
```

Comment:
use regular dict indexing here:

`data['recent_donation', amount]`

`.setdefault` is for when you want to use the result, and it might already be in there.